public/app/features/variables/adhoc/actions.ts

```diff
@@ -0,0 +1,164 @@
+    }
+
+    const index = variable.filters.findIndex(f => f.key === options.key && f.value === options.value);
+
+    if (index === -1) {
+      const { value, key, operator } = options;
+      const filter = { value, key, operator, condition: '' };
+      return await dispatch(addFilter(variable.uuid!, filter));
+    }
+
+    const filter = { ...variable.filters[index], operator: options.operator };
+    return await dispatch(changeFilter(variable.uuid!, { index, filter }));
+  };
+};
+
+export const changeFilter = (uuid: string, update: AdHocVariabelFilterUpdate): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterUpdated(toVariablePayload(variable, update)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const removeFilter = (uuid: string, index: number): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterRemoved(toVariablePayload(variable, index)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const addFilter = (uuid: string, filter: AdHocVariableFilter): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filterAdded(toVariablePayload(variable, filter)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const setFiltersFromUrl = (uuid: string, filters: AdHocVariableFilter[]): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const variable = getVariable(uuid, getState());
+    dispatch(filtersRestored(toVariablePayload(variable, filters)));
+    await dispatch(variableUpdated(toVariableIdentifier(variable), true));
+  };
+};
+
+export const changeVariableDatasource = (datasource: string): ThunkResult<void> => {
+  return async (dispatch, getState) => {
+    const { editor } = getState().templating;
+    const variable = getVariable(editor.id, getState());
+
+    const loadingText = 'Adhoc filters are applied automatically to all queries that target this datasource';
+
+    dispatch(
+      changeVariableEditorExtended({
+        propName: 'infoText',
+        propValue: loadingText,
+      })
+    );
+    dispatch(changeVariableProp(toVariablePayload(variable, { propName: 'datasource', propValue: datasource })));
+
+    const ds = await getDatasourceSrv().get(datasource);
+    console.log('ds', ds);
+
+    if (!ds || !ds.getTagKeys) {
+      dispatch(
+        changeVariableEditorExtended({
+          propName: 'infoText',
+          propValue: 'This datasource does not support adhoc filters yet.',
+        })
+      );
+    }
+  };
+};
+
+export const initAdHocVariableEditor = (): ThunkResult<void> => dispatch => {
+  const dataSources = getDatasourceSrv().getMetricSources();
```

Comment:
Doesn't getMetricSources return a Promise?
```suggestion
  const dataSources = await getDatasourceSrv().getMetricSources();
```

Reply:
I thought so as well (at first) but it doesn't look like it when I inspected the code.

Reply:
Ok, my bad